syntax = "proto3";

message CreateRequest {
  string filename = 1;
  string contents = 2;
}

message CreateDirectoryRequest {
  string directory = 1;
}

message UpdateRequest {
  string filename = 1;
  string contents = 2;
}

message GetDirectoryRequest {
  string directory = 1;
}

message GetFileRequest {
  string filename = 1;
}

message DeleteRequest {
  string filename = 1;
}

message StorageServiceRequest {
  int32 request_id = 1;
  string user = 2;
 
  oneof service_request {
    CreateRequest c = 3;
    CreateDirectoryRequest cd = 4;
    UpdateRequest u = 5;
    GetDirectoryRequest gd = 6;
    GetFileRequest gf = 7;
    DeleteRequest d = 8;
  };
};

enum ResponseCode {
  SUCCESS = 0;
  FAILURE = 1;
  INVALID = 2;
};

message GenericResponse {
};

message GetDirectoryResponse {
  string page = 1;
};

message GetFileResponse {
  string contents = 1;
};

message FailureResponse {
  string error_message = 1;
};

message StorageServiceResponse {
  int32 request_id = 1;
  string user = 2;
  
  ResponseCode response_code = 3;
  oneof service_response {
    GenericResponse g = 4;
    GetDirectoryResponse p = 5;
    GetFileResponse c = 6;
    FailureResponse failure = 7;
  };
};
